### ISSUE TITLE
Possibly replace the use of wait.Poll in e2e tests with some other standard retry mechanism?

### ISSUE THREAD

**Issue by satnam6502** [kubernetes/kubernetes#10307]
I have just had to re-write this loop:

``` go
    var got *api.Namespace
    for t := time.Now(); time.Since(t) < singleCallTimeout; time.Sleep(poll) {
        got, err = c.Namespaces().Create(namespaceObj)
        if err == nil {
            break
        }
    }
    if err != nil {
        return nil, err
    }
```

as:

``` go
    var got *api.Namespace
    if err := wait.Poll(poll, timeout, func() (bool, error) {
        var err error
        got, err = c.Namespaces().Create(namespaceObj)
        if err != nil {
            return false, nil
        }
        return true, nil
    }); err != nil {
        return nil, err
    }
```

because we are supposed to use `wait.Poll` in our tests.

I don't llike this. With the lack of generics in Go we can't do a good job of writing such a function. Ideally, we should paramaterize on result type of the operation so created values can be returned and so that any information required by the closure can be passed as an argument. We can't do this in Go today so in the closure we end up using a value from the outer scope (here `got`) which is horrible. I also don't like the use of a closure here (and the underlying implementation uses channels) -- all of this seems like a lot of complexity for what is a fairly simple procedure.

Personally, I find the `wait.Poll` pattern hard to understand and I think it makes our tests harder to write and maintain.

@quinton-hoole @ixdy @mbforbes

**Comment by mbforbes** [comment]
My main dislike of it is that the function it accepts has two return values instead of one for the sole purposing of short-circuiting out on a hard error.

A lot of the other Go code I am familiar with just has a single `error` as the return, which I think is a better interface. I have never actually used this to short-circuit out on a non-recoverable error, and I've used it several times now.

**Comment by satnam6502** [comment]
@mbforbes just wrote to me in IM:
```

thank you
btw there's a typo---you `return nil, err` but the first must be a bool
````
which confirms my point about how confusing this is.

**Comment by quinton-hoole** [comment]
@lavalamp FYI.  I think that he probably wrote wait.Poll() originally, and may be able to defend it's design better than I can.

I don't feel terribly strongly about how exactly we implement retries, only that we do it consistently and sensibly (e.g. timing out after a well defined time period where appropriate, and not after a number of iterations of an operation with poorly defined timing semantics, such that you don't actually know how long a retry loop will take).  

Waiting for go to implement generics is going to be like waiting for the cold death of the universe.  I've retitled this issue accordingly. :-)

**Comment by satnam6502** [comment]
The bit about waiting for generics was meant to be a joke.

**Comment by satnam6502** [comment]
Also, @mbforbes has told me that I have got my loop wrong because I am not correctly using the bool and error return values.

**Comment by lavalamp** [comment]
No, wait.Poll is @smarterclayton I'm pretty sure.

Many of the places we use wait.Poll can now be done better with watches. I would like to eventually move to that.

I personally don't find wait.Poll to make much impact on clarity, except that I always have to look up whether the timeout or poll period is the first argument.

**Comment by quinton-hoole** [comment]
That's what compilers are for :-)

**Comment by smarterclayton** [comment]
So much hate :) - the biggest benefit is that it prevents lots of small inconsistent places to be consistent.  I've seen people mess up basic loops and wait.Poll equally, but it does give us something to look at.

I would like to have wait.Watch() eventually (or an equiv) - the logic people do for watching doesn't often account for error cases properly.

**Comment by smarterclayton** [comment]
Maybe we should type the first argument.

**Comment by satnam6502** [comment]
Well, no hate intended! More comical frustration at my brain being too
small to understand `wait.Poll` :-)

**Comment by smarterclayton** [comment]
To be fair, I also can't remember which order they are in :)

Part of the reason for wait poll is that I wanted to do wait.Backoff and wait.ExponentialBackoff as well.

**Comment by satnam6502** [comment]
I would prefer a function which has a (not currently valid) type like this:

```
func RetryUntilTimeout(pollInterval time.Duration, timeout time.Duration, operation func(generic X) generic Y) (Y, error)
```
which would keep retrying the operation at the given frequency until the specified timeout.
The operation is passed an argument of type `X` so it has everything it needs in the closure to work without directly accessing the outer context and the returns the result of the operation of type Y (e.g. a pod or namespace etc.).

**Comment by satnam6502** [comment]
Anyway, I will close this issue since this is not the biggest thing we need to worry about right now.

**Comment by smarterclayton** [comment]
Retry is definitely a clearer name

**Comment by ixdy** [comment]
As I mentioned in https://github.com/GoogleCloudPlatform/kubernetes/pull/10301#issuecomment-117857163, we really should at least execute the TODO on https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/util/wait/wait.go#L49 and create a PollImmediate, since this will otherwise unnecessarily slow down tests.

**Comment by smarterclayton** [comment]
I think we can change the method directly - the reason it's the way it is
was simplicity, not a desire to wait the first interval.  Checking
immediately seems like something we'd always want.

On Jul 1, 2015, at 8:09 PM, Jeff Grafton <notifications@github.com> wrote:

As I mentioned in #10301 (comment)
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10301#issuecomment-117857163>,
we really should at least execute the TODO on
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/util/wait/wait.go#L49
and create a PollImmediate, since this will otherwise unnecessarily slow
down tests.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/10307#issuecomment-117858154>
.

**Comment by ixdy** [comment]
We probably don't want to

**Comment by quinton-hoole** [comment]
@thockin Just implemented a PollImmediate(), so we can move over to that on a case-by-case basis.

I think we can close this one now.

**Comment by thockin** [comment]
#14957 

I could just change Poll() or leave it as PollImmediate().  The test speedup is dramatic.

**Comment by quinton-hoole** [comment]
@thockin it would definitely be interesting to experiment with switching Poll() to PollImmediate()'s behavior.  But I'd wait for a time when e2e is running nice and green, because I would expect the switch to potentially cause multiple difficult to diagnose e2e regressions.  And don't merge the PR until it's got a good few (say 20-30) clean e2e runs under it's belt.

**Comment by quinton-hoole** [comment]
To be clear to others who were not part of the offline conversation, Poll() is not unly used in tests, but all over our system code base, so changing it to PollImmediate() would change all of that quite substantially.

**Comment by thockin** [comment]
Exactly - I'd like to get this PR in ASAP (because flakes and test speed)
and then consider just making Poll() be PollImmediate() as a second or
third order thing.

**Comment by smarterclayton** [comment]
We already did that a week ago and it got reverted because it made things
flaky.  Won't this just happen the same way?

https://github.com/kubernetes/kubernetes/pull/14365

On Oct 2, 2015, at 6:45 AM, Tim Hockin <notifications@github.com> wrote:

Exactly - I'd like to get this PR in ASAP (because flakes and test speed)
and then consider just making Poll() be PollImmediate() as a second or
third order thing.

On Thu, Oct 1, 2015 at 6:03 PM, Quinton Hoole <notifications@github.com>
wrote:

> To be clear to others who were not part of the offline conversation,
> Poll() is not unly used in tests, but all over our system code base, so
> changing it to PollImmediate() would change all of that quite
substantially.
>
> —
> Reply to this email directly or view it on GitHub
> <
https://github.com/kubernetes/kubernetes/issues/10307#issuecomment-144891350
>
> .
>

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/10307#issuecomment-144917401>
.

**Comment by thockin** [comment]
Different impl, same goal.  The change in test performance is pretty
noticeable - is the concern about flakes based on evidence or just
suspicion?

**Comment by smarterclayton** [comment]
You'll have to ask @erictune - I didn't see additional discussion or a
response in either issue

On Oct 2, 2015, at 8:33 AM, Tim Hockin <notifications@github.com> wrote:

Different impl, same goal. The change in test performance is pretty
noticeable - is the concern about flakes based on evidence or just
suspicion?

On Thu, Oct 1, 2015 at 11:25 PM, Clayton Coleman <notifications@github.com>
wrote:

> We already did that a week ago and it got reverted because it made things
> flaky. Won't this just happen the same way?
>
> https://github.com/kubernetes/kubernetes/pull/14365
>
> On Oct 2, 2015, at 6:45 AM, Tim Hockin <notifications@github.com> wrote:
>
> Exactly - I'd like to get this PR in ASAP (because flakes and test speed)
> and then consider just making Poll() be PollImmediate() as a second or
> third order thing.
>
> On Thu, Oct 1, 2015 at 6:03 PM, Quinton Hoole <notifications@github.com>
> wrote:
>
> > To be clear to others who were not part of the offline conversation,
> > Poll() is not unly used in tests, but all over our system code base, so
> > changing it to PollImmediate() would change all of that quite
> substantially.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
>
>
https://github.com/kubernetes/kubernetes/issues/10307#issuecomment-144891350
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <
>
https://github.com/kubernetes/kubernetes/issues/10307#issuecomment-144917401
> >
>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <
https://github.com/kubernetes/kubernetes/issues/10307#issuecomment-144935099
>
> .
>

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/10307#issuecomment-144935837>
.
